bin/sozo/src/commands/options/starknet.rs

```diff
@@ -24,10 +27,22 @@ impl StarknetOptions {
     pub fn provider(
         &self,
         env_metadata: Option<&Environment>,
+        request_timeout_ms: Option<u64>,
     ) -> Result<(JsonRpcClient<HttpTransport>, String)> {
         let url = self.url(env_metadata)?;
-        trace!(?url, "Creating JsonRpcClient with given RPC URL.");
-        Ok((JsonRpcClient::new(HttpTransport::new(url.clone())), url.to_string()))
+        trace!(?url, "Creating JsonRpcClient with given RPC URL and timeout.");
+
+        let client = if let Some(request_timeout_ms) = request_timeout_ms {
+            ClientBuilder::default()
+                .timeout(Duration::from_millis(request_timeout_ms))
+                .build()
+                .unwrap()
+        } else {
+            ClientBuilder::default().build().unwrap()
+        };
+
+        let transport = HttpTransport::new_with_client(url.clone(), client);
+        Ok((JsonRpcClient::new(transport), url.to_string()))
```

Comment:
I'm just wondering at which poing a declaration can last based on the network + size of a class. This is the main reason why I exposed it.

But I do agree, if we can find something that could be handled without actually exposing it, it's definitely better I agree.
Retry once with bigger timeout and fail afterward perhaps?

Reply:
Hmmm good point. 

I think if we do retry, we'd just be sending the same request twice. In the case of class declaration, the timeout will be caused by the server processing the request too slow. And because the timeout is happening on the client, the server would still be processing the request, no?

Reply:
> Hmmm good point.
> 
> I think if we do retry, we'd just be sending the same request twice. In the case of class declaration, the timeout will be caused by the server processing the request too slow. And because the timeout is happening on the client, the server would still be processing the request, no?

Yep, that's a good point. Which may lead to undesired effect.
So I guess it's preferable to have a longer timeout, but may be shortened manually then?

Reply:
not exactly sure... a longer timeout (something like 20s) is probably a good number. if a request takes more than 20s, and if its not due to super slow machine, then there's probably something wrong in how katana process the request.

katana also times out at 20s.

https://github.com/dojoengine/dojo/blob/712422fc58b87aa1782220151ba33c057e4313e5/crates/katana/node/src/lib.rs#L308-L311